mods/logon-logoff-shutdown-sounds.wh.cpp

```diff
@@ -0,0 +1,582 @@
+{
+    HRESULT hr = S_OK;
+    static WCHAR wszSessionKey[256];
+    LONG Error;
+ 
+    if (DWORD dwSessionId = GetCurrentSessionId())
+    {
+        swprintf(wszSessionKey,
+                      L"SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Explorer\\SessionInfo\\%d",
+                      dwSessionId);
+    }
+    else
+        hr = HRESULT_FROM_WIN32(GetLastError());
+ 
+    if(SUCCEEDED(hr))
+    {
+        Error = RegCreateKeyExW(HKEY_LOCAL_MACHINE, wszSessionKey, 0, NULL,
+                                REG_OPTION_VOLATILE, samDesired, NULL, phKey, NULL);
+        if (Error != ERROR_SUCCESS)
+            hr = HRESULT_FROM_WIN32(Error);
+    }
+    
+    return hr;
+}
+
+//  a three state boolean for bools that need initialization
+typedef enum 
+{
+    TRIBIT_UNDEFINED = 0,
+    TRIBIT_TRUE,
+    TRIBIT_FALSE,
+} TRIBIT;
+
+TRIBIT g_tbPlayedStartupSound;
+DWORD g_dwCurProcId;
+
+BOOL WINAPI HasPerLogonActionBeenDone(LPCWSTR lpSubKey, TRIBIT* out_tbDone)
+{
+    DWORD dwDisposition;
+    HKEY hKey;
+    HKEY phkResult;
+
+    *out_tbDone = TRIBIT_TRUE;
+    if ( (int)SHCreateSessionKey(131078, &hKey) >= 0 )
+    {
+        if ( !RegCreateKeyExW(hKey, lpSubKey, 0, 0, 1u, 0x20006u, 0, &phkResult, &dwDisposition) )
+        {
+            RegCloseKey(phkResult);
+            if ( dwDisposition == 1 )
+                *out_tbDone = TRIBIT_FALSE;
+        }
+        RegCloseKey(hKey);
+    }
+    return *out_tbDone == TRIBIT_TRUE;
+}
+
+BOOL WINAPI HasLogonSoundBeenPlayed()
+{
+    return HasPerLogonActionBeenDone(L"LogonSoundHasBeenPlayed", &g_tbPlayedStartupSound);
+}
+
+BOOL WINAPI IsDesktopProcess() {
+    DWORD dwShellProcId = NULL;
+    DWORD dwCurProcId = GetCurrentProcessId();
+    HWND hShellWindow = FindWindow(TEXT("Shell_TrayWnd"), NULL);
+    if (!hShellWindow) // no desktop present yet, assume it's main explorer starting up
+    {
+        g_dwCurProcId = dwCurProcId;
+        return TRUE;
+    }
+    GetWindowThreadProcessId(hShellWindow, &dwShellProcId);
+    if (dwCurProcId == dwShellProcId || !hShellWindow) {
+        if (dwCurProcId != g_dwCurProcId) // because the dll handle is shared, the boolean needs to be reset when the explorer process changes
+        {
+            g_tbPlayedStartupSound = TRIBIT_UNDEFINED;
+            g_dwCurProcId = dwCurProcId;
+        }
+        return TRUE;
+    } else {
+        g_dwCurProcId = dwCurProcId;
+        return FALSE;
+    }
+}
+
+// Load the settings for the mod
+void LoadSettings(void)
+{
+    settings.bXpMode = (BOOL)Wh_GetIntSetting(L"xpmode");
+}
+
+// The mod is being initialized, load settings, hook functions, and do other
+// initialization stuff if required.
+BOOL Wh_ModInit() {
+    Wh_Log(L"Init");
+
+    LoadSettings();
+
+    g_hExplorer = GetModuleHandleW(NULL);
+    if (!g_hExplorer)
+        Wh_Log(L"Failed to get Explorer's handle");
+
+    g_hShlwapi = LoadLibrary(TEXT("shlwapi.dll"));
+
+    WindhawkUtils::SYMBOL_HOOK aHooks[] = {
```

Comment:
As the message says, rename the hooks variable. In this case, rename it to `explorerExeHooks` or one of the other accepted variants.